src/main/scala/chisel3/UnitTest.scala

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.BaseModule
+import chisel3.experimental.hierarchy.Definition
+import java.io.File
+import java.util.jar.JarFile
+import scala.collection.JavaConverters._
+
+/** All classes and objects marked as `UnitTest` are automatically discoverable
+ *  by the `DiscoverUnitTests` helper. */
+trait UnitTest
+
+/** Helper to discover all subtypes of `UnitTest` in the class path, and call
+ *  their constructors (if they are a class) or ensure that the singleton is
+ *  constructed (if they are an object).
+ *
+ *  This code is loosely based on the test suite discovery in scalatest, which
+ *  performs the same scan over the classpath JAR files and directories, and
+ *  guesses class names based on the encountered directory structure. */
+object DiscoverUnitTests {
+
+  /** The callback invoked for each unit test class name and unit test
+   *  constructor. */
+  type Callback = (String, () => Unit) => Unit
+
+  /** Discover all tests in the classpath and call `cb` for each. */
+  def apply(cb: Callback): Unit = classpath().foreach(discoverFile(_, cb))
+
+  /** Return the a sequence of files or directories on the classpath. */
+  private def classpath(): Iterable[File] = System
+    .getProperty("java.class.path")
+    .split(File.pathSeparator)
+    .map(s => if (s.trim.length == 0) "." else s)
+    .map(new File(_))
+
+  /** Discover all tests in a given file. If this is a JAR file, looks through
+   *  its contents and tries to find its classes. */
+  private def discoverFile(file: File, cb: Callback): Unit = {
+    if (file.getPath.toLowerCase.endsWith(".jar")) {
+      val jarFile = new java.util.jar.JarFile(file)
+      jarFile.entries.asScala.foreach { jarEntry =>
+        val name = jarEntry.getName
+        if (!jarEntry.isDirectory && name.endsWith(".class"))
+          discoverClass(pathToClassName(name), cb)
+      }
+    } else if (file.isDirectory) {
+      def visit(prefix: String, file: File): Unit = {
+        val name = prefix + "/" + file.getName
+        if (file.isDirectory) {
+          for (entry <- file.listFiles)
+            visit(name, entry)
+        } else if (name.endsWith(".class")) {
+          discoverClass(pathToClassName(name), cb)
+        }
+      }
+      for (entry <- file.listFiles)
+        visit("", entry)
+    }
+  }
```

Comment:
Nit: I'd opt for either: (1) early return style or (2) a match. I.e.,:

``` scala
if (file.getPath.toLowerCase.endsWith(".jar")) {
  /** ... */
  return
}

if (!file.isDirectory) 
  return

/** Handle directory */
```

Or:

``` scala
file match {
  case jarFile if jarFile.getPath.toLowerCase.endsWith(".jar") => /** handle jar filr */
  case dir if dir.isDirectory => /** handle directory */
  case _ => 
}
```

Most Scala code tends towards (2) as it is more obvious what is going on.

Reply:
Other piece: both `discoverFile` and `visit` look to be tail recursive. You can force the compiler to check that this is definitely the case with:

``` scala
import scala.annotation

@annotation.tailrec
def foo(): Unit = ???
```

Reply:
The case style looks pretty readable :+1: Tail recursion doesn't work unfortunately because most recursive calls are inside loops.